Add v24.1.2 Cloud release notes (#18699)
cockroachdb/docs: src/current/_includes/releases/v24.1/v24.1.2.md

## v24.1.2

Release Date: July 2, 2024

{% include releases/new-release-downloads-docker-image.md release=include.release %}

<h3 id="v24-1-2-{{-site.data.products.enterprise-}}-edition-changes">{{ site.data.products.enterprise }} edition changes</h3>

> - Fixed a bug in CockroachDB v22.2 and later where long running initial scans may incorrectly restore [checkpoint job](https://www.cockroachlabs.com/docs/v24.1/how-does-an-enterprise-changefeed-work) progress and drop events during node (and thus [changefeed](https://www.cockroachlabs.com/docs/v24.1/create-changefeed)) restart. This issue was most likely to occur in clusters where (1) the `changefeed.shutdown_checkpoint.enabled` [cluster setting](https://www.cockroachlabs.com/docs/v24.1/cluster-settings) is `true`, (2) there are multiple table targets in a changefeed, or (3) low values of the cluster settings `changefeed.frontier_checkpoint_frequency` or [`changefeed.frontier_highwater_lag_checkpoint_threshold`](https://www.cockroachlabs.com/docs/v24.1/cluster-settings#setting-changefeed-frontier-highwater-lag-checkpoint-threshold) are in effect. [#124996][#124996]

Author: Thanks that's really helpful!  I've updated this to use the much better version from that file.

- Added an error message for unsupported Apache Pulsar [changefeed](https://www.cockroachlabs.com/docs/v24.1/create-changefeed) sink parameters. [#124666][#124666]

<h3 id="v24-1-2-operational-changes">Operational changes</h3>

- Improved [disk usage metric reporting](https://www.cockroachlabs.com/docs/v24.1/ui-cluster-overview-page#capacity-metrics) over volumes that dynamically change their size over the life of the `cockroach` process. [#125050][#125050]

> - The following [cluster settings](https://www.cockroachlabs.com/docs/v24.1/cluster-settings) have had their default values updated from `100000` to `7500`. These settings control the maximum number of [fingerprints](https://www.cockroachlabs.com/docs/v24.1/ui-statements-page#sql-statement-fingerprints) (distinct combinations of statements and [transactions](https://www.cockroachlabs.com/docs/v24.1/transactions)) we can store in memory, with stats being flushed on a `10m` [interval](https://www.cockroachlabs.com/docs/v24.1/interval) by default. Users with workloads producing unique fingerprints higher than this amount within the flush interval may notice that they are missing SQL stats. Users who want to capture those missing stats may increase these settings as necessary. [#125554][#125554]

Reviewer: ```suggestion
- The default values for the following [cluster settings](https://www.cockroachlabs.com/docs/v24.1/cluster-settings) were updated from `100000` to `7500`. These settings control the maximum number of [fingerprints](https://www.cockroachlabs.com/docs/v24.1/ui-statements-page#sql-statement-fingerprints) (distinct combinations of statements or [transactions](https://www.cockroachlabs.com/docs/v24.1/transactions)) CockroachDB can store in memory, with stats being flushed on a `10m` [interval](https://www.cockroachlabs.com/docs/v24.1/interval) by default. You can increase these settings if your workload produces more unique fingerprints than this amount within the flush interval, and you notice that SQL stats are missing. [#125554][#125554]
```

Author: Thanks! I have slurped in this whole thing except that the release note commit message in cockroachdb/cockroach#125554 said "statements _and_ transactions" so I'm leaving that term

  - [`sql.metrics.max_mem_stmt_fingerprints`](https://www.cockroachlabs.com/docs/v24.1/cluster-settings#setting-sql-metrics-max-mem-stmt-fingerprints)
  - [`sql.metrics.max_mem_txn_fingerprints`](https://www.cockroachlabs.com/docs/v24.1/cluster-settings#setting-sql-metrics-max-mem-txn-fingerprints)

<h3 id="v24-1-2-db-console-changes">[DB Console](https://www.cockroachlabs.com/docs/v24.1/ui-overview) changes</h3>

- The favicon now renders properly for [DB Console](https://www.cockroachlabs.com/docs/v24.1/ui-overview), along with other image files. [#122706][#122706]

<h3 id="v24-1-2-bug-fixes">Bug fixes</h3>

> - [`SHOW TYPES`](https://www.cockroachlabs.com/docs/v24.1/show-types) now includes [user defined composite types](https://www.cockroachlabs.com/docs/v24.1/create-type#create-a-composite-data-type). It had omitted those types ever since composite types were added in v23.1. [#124817][#124817]

Reviewer: ```suggestion
- [`SHOW TYPES`](https://www.cockroachlabs.com/docs/v24.1/show-types) now includes [user-defined composite types](https://www.cockroachlabs.com/docs/v24.1/create-type#create-a-composite-data-type). It had omitted those types ever since composite types were added in v23.1. [#124817][#124817]
```

Author: FINE THEN RYANNNNNNNN 🤣 

ps thank you :-)

> - Fixed a crash introduced in v24.1.0-beta.2 that could occur when planning [statistics](https://www.cockroachlabs.com/docs/v24.1/cost-based-optimizer#table-statistics) collection on a table with a [virtual computed column](https://www.cockroachlabs.com/docs/v24.1/computed-columns) using a [user-defined type](https://www.cockroachlabs.com/docs/v24.1/create-type) when the newly-introduced [cluster setting](https://www.cockroachlabs.com/docs/v24.1/cluster-settings#setting-sql-stats-virtual-computed-columns-enabled) `sql.stats.virtual_computed_columns.enabled` is set to `true`. (The setting was introduced in v24.1.0-alpha.1.) [#124996][#124996]

Reviewer: ```suggestion
- Fixed a crash introduced in v24.1.0-beta.2 that could occur when planning [statistics](https://www.cockroachlabs.com/docs/v24.1/cost-based-optimizer#table-statistics) collection on a table with a [virtual computed column](https://www.cockroachlabs.com/docs/v24.1/computed-columns) using a [user-defined type](https://www.cockroachlabs.com/docs/v24.1/create-type) when the [cluster setting](https://www.cockroachlabs.com/docs/v24.1/cluster-settings#setting-sql-stats-virtual-computed-columns-enabled) `sql.stats.virtual_computed_columns.enabled` (introduced in v24.1.0-alpha.1) was set to `true`. [#124996][#124996]
```

Author: Oooh better thanks!

> - Fixed handling in the [declarative schema changer](https://www.cockroachlabs.com/docs/v24.1/online-schema-changes#declarative-schema-changer) when columns are included in the `STORING()` clause of a [`CREATE INDEX`](https://www.cockroachlabs.com/docs/v24.1/create-index) statement. We now check if the column is virtual up front, and properly detect when a column is already handled by an existing [`INDEX`](https://www.cockroachlabs.com/docs/v24.1/indexes) when the column name has UTF-8 characters. [#125153][#125153]

Reviewer: ```suggestion
- Fixed handling in the [declarative schema changer](https://www.cockroachlabs.com/docs/v24.1/online-schema-changes#declarative-schema-changer) when columns are included in the `STORING()` clause of a [`CREATE INDEX`](https://www.cockroachlabs.com/docs/v24.1/create-index) statement. CockroachDB now checks if the column is virtual up-front, and properly detects when a column is already handled by an existing [`INDEX`](https://www.cockroachlabs.com/docs/v24.1/indexes) when the column name has UTF-8 characters. [#125153][#125153]
```

Author: Snagged it!

- Fixed a bug where [telemetry logs](https://www.cockroachlabs.com/docs/v24.1/logging#telemetry) had the same [statement fingerprint](https://www.cockroachlabs.com/docs/v24.1/ui-statements-page#sql-statement-fingerprints) ID for different SQL statements. [#125000][#125000]
- Fixed an issue where [adding a column](https://www.cockroachlabs.com/docs/v24.1/alter-table#add-column) with a default value of an empty [array](https://www.cockroachlabs.com/docs/v24.1/array) would not succeed. [#125325][#125325]
- [`ALTER TABLE ... ADD CONSTRAINT UNIQUE`](https://www.cockroachlabs.com/docs/v24.1/alter-table) will now fail with a well-formed error message and code `42601` if a statement tries to add a [unique constraint](https://www.cockroachlabs.com/docs/v24.1/unique) on an expression. [#125417][#125417]

> - Fixed a bug where the [`public` schema](https://www.cockroachlabs.com/docs/v24.1/schema-design-overview#schemas) would be created with the wrong owner. Previously the [`admin` role](https://www.cockroachlabs.com/docs/v24.1/security-reference/authorization#roles) would own the `public` schema. Now, the database owner is also the owner of the `public` schema. The ownership can be altered after the schema is created. [#125533][#125533]

Reviewer: ```suggestion
- Fixed a bug where the [`public` schema](https://www.cockroachlabs.com/docs/v24.1/schema-design-overview#schemas) would be created with the wrong owner. Previously the [`admin` role](https://www.cockroachlabs.com/docs/v24.1/security-reference/authorization#roles) would own the `public` schema. Now, the database owner is also the owner of the `public` schema. The owner can be altered after the schema is created. [#125533][#125533]
```

Author: ✅

- Fixed a bug in v24.1, v23.2, and v23.1 where using the `changefeed.aggregator.flush_jitter` [cluster setting](https://www.cockroachlabs.com/docs/v24.1/cluster-settings#setting-changefeed-aggregator-flush-jitter) with the [`min_checkpoint_frequency`](https://www.cockroachlabs.com/docs/v24.1/create-changefeed#min-checkpoint-frequency) option set to `0` could cause panics. [#125459][#125459]

> - Resolved a concerning log message that said `"expiration of liveness record ... is not greater than expiration of the previous lease ... after liveness heartbeat"`. This message is no longer possible. [#125449][#125449]

Reviewer: ```suggestion
- The log message `"expiration of liveness record ... is not greater than expiration of the previous lease ... after liveness heartbeat"` is no longer generated. [#125449][#125449]
```

Author: yes that language seems much less concern-inducing :-)

> - Previously, CockroachDB would hit an internal error when evaluating [`INSERT`s](https://www.cockroachlabs.com/docs/v24.1/insert) into [`REGIONAL BY ROW`](https://www.cockroachlabs.com/docs/v24.1/set-locality#set-the-table-locality-to-regional-by-row) tables where the source is a `VALUES` clause with a single row and at least one boolean expression; this is now fixed. The bug was introduced in the v23.2.0 release. [#126209][#126209]

Reviewer: ```suggestion
- Fixed a bug introduced in v23.2.0 in which CockroachDB would hit an internal error when evaluating [`INSERT`s](https://www.cockroachlabs.com/docs/v24.1/insert) into [`REGIONAL BY ROW`](https://www.cockroachlabs.com/docs/v24.1/set-locality#set-the-table-locality-to-regional-by-row) tables where the source was a `VALUES` clause with a single row and at least one boolean expression. [#126209][#126209]
```

Author: ✅

<h3 id="v24-1-2-performance-improvements">Performance improvements</h3>

> - Make error handling in the [vectorized execution](https://www.cockroachlabs.com/docs/v24.1/vectorized-execution) engine much cheaper. This should help avoid bad metastable regimes perpetuated by statement timeout handling consuming all CPU time, leading to more statement timeouts. [#124996][#124996]

Reviewer: I suggest reusing the version of the note published in `v23.2.6.md` as it's clearer.

Author: nice, snagged it

> - Some [privilege](https://www.cockroachlabs.com/docs/v24.1/security-reference/authorization#managing-privileges) checks when scanning the `crdb_internal.system_jobs` internal table now happen once before the scan, instead of once for each row. This will improve performance for queries that read from `crdb_internal.system_jobs`. [#125250][#125250]

Reviewer: ```suggestion
- Some [privilege](https://www.cockroachlabs.com/docs/v24.1/security-reference/authorization#managing-privileges) checks when scanning the `crdb_internal.system_jobs` internal table now happen once before the scan, instead of once for each row. This improves performance for queries that read from `crdb_internal.system_jobs`. [#125250][#125250]
```

Author: ✅

<h3 id="v24-1-2-miscellaneous">Miscellaneous</h3>

> - Disabled a [changefeed](https://www.cockroachlabs.com/docs/v24.1/create-changefeed) optimization on reducing duplicates during aggregator restarts due to its bad performance. [#124996][#124996]

Reviewer: This should be moved under Operational changes (see `v23.2.5`).

```suggestion
- Disabled a [changefeed](https://www.cockroachlabs.com/docs/v24.1/create-changefeed) optimization to reduce duplicates during aggregator restarts has been disabled due to poor performance. [#124996][#124996]
```

Author: Split the difference and went with 

> - A [changefeed](https://www.cockroachlabs.com/docs/v24.1/create-changefeed) optimization to reduce duplicates during aggregator restarts has been disabled due to poor performance. [#124996][#124996]
